code/datums/components/material_container.dm

```diff
@@ -10,42 +10,79 @@
 */
 
 /datum/component/material_container
+	/// The total amount of materials this material container contains
 	var/total_amount = 0
+	/// The maximum amount of materials this material container can contain
 	var/max_amount
+	/// Unused?
 	var/sheet_type
+	/// Map of material ref -> amount
 	var/list/materials //Map of key = material ref | Value = amount
+	/// Unused?
 	var/disable_attackby
-	var/list/allowed_typecache
+	/// The list of materials that this material container can accept
+	var/list/allowed_materials
+	/// The typecache of things that this material container can accept
+	var/list/allowed_item_typecache
+	/// The last main material that was inserted into this container
 	var/last_inserted_id
+	/// Whether or not this material container allows specific amounts from sheets to be inserted
 	var/precise_insertion = FALSE
+	/// A callback for checking wheter we can insert a material into this container
+	var/datum/callback/insertion_check
+	/// A callback invoked before materials are inserted into this container
 	var/datum/callback/precondition
+	/// A callback invoked after materials are inserted into this container
 	var/datum/callback/after_insert
-	///The material container flags. See __DEFINES/materials.dm.
+	/// The material container flags. See __DEFINES/materials.dm.
 	var/mat_container_flags
 
 /// Sets up the proper signals and fills the list of materials with the appropriate references.
-/datum/component/material_container/Initialize(list/mat_list, max_amt = 0, _mat_container_flags=NONE, list/allowed_types, datum/callback/_precondition, datum/callback/_after_insert)
+/datum/component/material_container/Initialize(list/init_mats, max_amt = 0, _mat_container_flags=NONE, list/allowed_mats=init_mats, list/allowed_items, datum/callback/_insertion_check, datum/callback/_precondition, datum/callback/_after_insert)
 	materials = list()
 	max_amount = max(0, max_amt)
 	mat_container_flags = _mat_container_flags
 
-	if(allowed_types)
-		if(ispath(allowed_types) && allowed_types == /obj/item/stack)
-			allowed_typecache = GLOB.typecache_stack
+	allowed_materials = allowed_mats || list()
```

Comment:
I've said before that I don't really like the overuse of this operator, but this isn't exactly easy to read in the first place, so why not